### ISSUE TITLE
For Android with Float width, snapToOffset, snapToInterval is incorrect in the flash list

### ISSUE THREAD

**Issue by Siby1lA** [Shopify/flash-list#1153]
## Current behavior

When scrolling through snapToOffsets or snapToInterval for a horizontally rendered FlashList on Android with a floating screen width, the alignment is not consistent.

https://github.com/Shopify/flash-list/assets/77488652/8d594eba-239b-4a21-9579-ed395a78a173

## Expected behavior

In contrast, the iOS maintains consistent alignment.

https://github.com/Shopify/flash-list/assets/77488652/9029d2b5-e66d-4227-9c38-5a3d5c19c1ed

## To Reproduce

example on the snack : https://snack.expo.dev/@twi1ight/flashlist?platform=android

## Platform:

- [ ] iOS
- [x] Android

## Environment

react native v0.73.2 -> flashList v1.6.4
or
expo v50.0.4 -> flashList v1.6.3

**Comment by renochew** [comment]
Hi, I just want to chime in as I am also struggling with snapToOffset at the moment, the snap position keep shifting when using a long list. I have a very simple snack to show the problem and hope that would help developer to track down the issue.

[https://snack.expo.dev/@renochew/flashlist?platform=android](url)

In this snack I have 2 flashlists side by side, both have a stack of square view of 50 pixels, snapping the scrollview in a multiple of 50 should ideally have both list matching at exactly the same point. But that is not what happened at the moment.

Thanks.

**Comment by hmust92** [comment]
I am running into the EXACT same issue here.

**Comment by AspireOne** [comment]
Also experiencing this. Pretty huge issue for me.

My current workaround is to add an offset of +0.2 to the number you would normally pass to snapToInterval.

```tsx
<FlashList
  ref={flashListRef}
  data={activities}
  renderItem={renderActivityCard}
  horizontal
  showsHorizontalScrollIndicator={false}
  estimatedItemSize={cardWidth} // without margin
  estimatedListSize={{
    width: screenWidth,
    height: cardHeight,
  }}
  // Hardcoded 0.2 offset.
  snapToInterval={cardWidth + cardMargin + 0.2}
  snapToAlignment={"start"}
  decelerationRate={0.8}
  disableIntervalMomentum
  onEndReachedThreshold={0.5}
  viewabilityConfig={viewabilityConfig}
  onViewableItemsChanged={onViewableItemsChanged}
  {...props.flashListProps}
/>
```

It *_almost_* balances it, and the drift starts to be apparent only after 100 items or so. I would have balanced it completely (an offset of +0.19 or +0.18 would be perfect), but FlashList seems to round it UP to 0.2 for 0.19, and DOWN to 0.1 for 0.18.

**Comment by muhammadSharaf** [comment]
facing the same issue, except for backward scrolling, it works fine. when I start to scroll forward, after some scrolls, it begins to shift.
Also I can confirm that FlatList works well with me with getItemLayout

**Comment by alextbogdanov** [comment]
This fixes it for me:
`
<FlashList
  ref={flashListRef}
  data={[...usersData, usersData[0]]}
  horizontal={false}
  onEndReachedThreshold={Platform.OS === 'android' ? 0.01 : undefined}
  onEndReached={handleEndReached}
  showsVerticalScrollIndicator={false}
  scrollEnabled={profilePicture1 ? true : false}
  pagingEnabled={Platform.OS === 'ios'}
  snapToAlignment={Platform.OS === 'ios' ? 'start' : 'end'}
  estimatedItemSize={
    screenHeight -
    bottomTabBarHeight +
    (Platform.OS === 'android' ? notchHeight : 0)
  }
  snapToInterval={
    Platform.OS === 'android'
      ? screenHeight - bottomTabBarHeight + notchHeight
      : undefined
  }
  decelerationRate={Platform.OS === 'android' ? 'normal' : undefined}
  keyExtractor={(item, index) => item.userID + index}
  disableIntervalMomentum={Platform.OS === 'android'}
  bounces={false}
  renderItem={({ item, index }) => (
    <ConnectUserCard
      item={item}
      index={index}
      length={usersData.length}
      loaded={loaded}
      key={item.userID + index}
    />
  )}
/>
`

**Comment by alextbogdanov** [comment]
This fixes it for me:

<img width="692" alt="Image" src="https://github.com/user-attachments/assets/588a391c-1349-49ff-8b77-1df532a3353f" />

**Comment by sergey-g-s** [comment]
I’m running into the same problem on Android when using snapToOffsets or snapToInterval in FlashList. Despite correctly calculating offsets, the snapping feels inconsistent—especially with float-based widths and margins.

Has anyone found a reliable solution or workaround for this?
Would appreciate any updates or insights

**Comment by Iuriy-Budnikov** [comment]
That hack helped to resolve an issue on android

```
const ANDROID_SNAP_ADJUSTMENT = Platform.OS === 'android' ? 0.2 : 0;
const PAGING_ENABLED = Platform.OS === 'ios';
const SNAP_TO_INTERVAL = MAX_CARD_WIDTH + LIST_GAP * 2 + ANDROID_SNAP_ADJUSTMENT;
const DECELERATION_RATE = Platform.OS === 'android' ? 0.99 : 'fast';
const DISABLE_INTERVAL_MOMENTUM = Platform.OS === 'android';
```

**Comment by NX-Michael** [comment]
I'm facing the same issue for this except, that I'm doing a vertical scroll instead and found my weird workaround so I'll share my side. 

### TLDR at the bottom

For my use case, each `renderScene` item's height is the full height of the FlashList and `snapToInterval` will also be the `FlashList` height. The issue I'm facing is that as I scroll, I'm seeing the previous item peeking in. And as I keep scrolling, more of the previous item keeps peeking down. 

How I'm doing it on my side is basically rendering the FlashList and getting its height, let's call it `H`, using `onLayout`, and setting it as a state to be used for the `snapToInverval` and `renderItem` components' height.

Played around with the values, found my own workaround. 

When `H` is a decimal number, all items rendered from the `renderItem` components will also follow the value `H`, but `snapToInterval` seems to be scrolling less than it should, and thus, with every scroll, the previous component is slowly peeking down from the top. I suspect either `snapToInterval` is not respecting float values, miscalulating, or `FlastList` is just giving false value for its height.

To try and fix this, I did a `Math.ceil` on `H` before setting the state and passing it to `snapToInterval` and the `renderScene` function. But the issue still persists. Checking the rendered item's height after the `Math.ceil`, it actually rendered larger than the rounded up value.

So let's say `H` was `176.xxxxx`. Without rounding up, all rendered items seem to respect this value and showed to have the height of `176.xxx`, but the scroll issue is there. After rounding to `177`, the rendered Item should be `177`, but actually rendered as `177.238xxx`. Notice the `0.2xxx` difference here and how everyone's patch is to add `0.2`. So there is now a misalign between the `snapToInterval` being `177` and the renderedItem being `177.238xxx`

So after this, I decided to add another state, lets call it `H2` which is the increased size from of `177.238xxx` and is what `snapToInterval` will be using. And this resolved my scrolling issue.

I still can't pinpoint the exact cause of this issue because each step of my workaround contradicts my suspicion.
- I suspected that either the FlashList is giving smaller values than its actual full height, but this doesn't make sense also because if I don't round up the height, both the FlashList and all the rendered Item will the same height. And I don't see the next story peeking in from the bottom. 
- Unless somehow both the FlashList and rendered items' height are returning lower than they are. But this also makes no sense because of the workaround I did gave me a larger height from the `Math.ceil`, so it seems to be accurate. 
- It can't be `snapToInterval` not following decimals, because it followed the increased renderItem height, which is a decimal.

### TLDR:
My suspicion of random issue made me found a weird workaround but can't pinpoint the cause.
- Round my initial floating point value to a whole number using `Math.ceil`
- Pass the rounded value to rendered Item
- Get new increased item height and used this floating point value as the `snapToInterval`

Example workaround:
```
  const [flashListHeight, setFlashListHeight] = useState(0);
  const [flashListItemHeight, setFlashListItemHeight] = useState(0);

...

  const renderItem = useCallback(
    () => {
      return (
        <View
          style={{
            height: flashListHeight,
            width: "100%",
          }}
          onLayout={(event) => {
            const { height } = event.nativeEvent.layout;
            setFlashListItemHeight(height);
          }}
        >
          ...
        </View>
      );
    },[ ]);

...

      <FlashList
        ...
        snapToInterval={flashListItemHeight}
        onLayout={(event) => {
          const { height } = event.nativeEvent.layout;
          setFlashListHeight(Math.ceil(height));
        }}
        ...
        />
```

**Comment by naqvitalha** [comment]
Hi! 👋

We're now focusing exclusively on **FlashList v2** and **React Native's New Architecture**. This means we're closing older issues and suggestions that aren't relevant to v2.

**If this issue/suggestion is still relevant:**
- Please verify it still occurs in FlashList v2
- Open a new issue with v2-specific details
- We'll be happy to take a look!

**Resources:**
- [Migration to v2](https://shopify.github.io/flash-list/docs/migration-guide)
- [FlashList Documentation](https://shopify.github.io/flash-list/)

Thank you for your understanding! 🙏
